src/System.Text.ValueBuilder/System/Text/ValueStringBuilder.Format.cs

```diff
@@ -0,0 +1,350 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Text
+{
+    public ref partial struct ValueStringBuilder
```

Comment:
I'm wondering if this strategy, which allows for rent/return, could be blended in with the `params` proposal at https://github.com/dotnet/csharplang/issues/2302 as well. Right now, ignoring cheap stack allocation for arrays (which put limitations on the size), a possible translation of:

```csharp
Use(1);
Use(1, 2);
Use(1, 2, 3);
```

with a

```csharp
void Use(params Span<int> xs) { ... }
```

could allocate an `int[3]` array, with `Span<int>` views over it for subsequent invocations of `Use`. However, it does seem to preclude library assistance to mediate the `int[]` allocation, so the optimization works well within a single method but not across. With the work in `System.Memory` on pools, it'd seem possible to rent/return arrays.

One way would be to simply substitute `newarr` for a `Rent` operation, provided the compiler has proven that the arrays can't escape (e.g. only spans are handed out, and no refs to elements therein are returned). That'd effectively result in a single `Rent(n)` operation, where `n` is the maximum of all lengths required in the body of a method. The corresponding `Return` would have to end up in a `try...finally...` construct. A drawback is the tight coupling of the language to another API. Another possible concern is a long-running method where `n` is big because of one single large `params` site, thus holding on to the large array for longer than necessary. FWIW, this is no different than lifetime issues with big closures (due to the union of all hoisted variables ending up in the same closure object, thus any delegate extends the lifetime of the entire closure), and `n` is likely never very big because it's proportional to the argument count in user-written code.

An alternative would be for each `params` site to result in a builder pattern, derived from the type of the parameter to which `params` is applied. Given such a type `P`, there's a builder type `B` which has the following operations:

* `static I Create(int capacity)` to create the intermediate object `I` (call that `i`),
* `i.Add(T)` or `i.Set(int, T)` methods to add or set elements. These could return `void` or `I` (to create a fluent pattern) and are bound with maximum flexibility (e.g. allowing extension methods).
   * Note I'm omitting `T this[int] { set; }` as an option for a member on `i` because it'd only work if all elements are of the same type (which brings me to interpolated string further on).
* `i.Result` is a property or field of type `P`, returning the result, and,
* `i` has an optional `Dispose`, e.g. for value types akin to `ValueStringBuilder` which can perform a `Return`.

This is somewhat similar to collection initializers simply looking for an `Add` method that can be called as an instance method. Also, looking up a builder type for a given type could be done in ways similar to task-like types (but supporting any generic arity by pouring generic arguments over between `P` and `B`).

For example:

```csharp
void Use(double d, params P xs) { ... }

Use(Math.Sin(x), 1, "foo", true);
```

would be syntactic sugar for:

```csharp
var __e = Math.Sin(x);

using (var __t = B.Create(3))
{
  __t.Set(0, 1);
  __t.Set(1, "foo");
  __t.Set(2, true);

  Use(__e, __t.Result);
}
```

with (pattern-based) `using` being optional, depending on whether `__t`'s type supports `Dispose`. This is given a definition of `P` like this:

```csharp
[CollectionBuilder(typeof(B))]
class P {}

static class B
{
  public static I Create(int capacity);

  public struct I
  {
    // Or a generic Set<T> would work, too.
    public void Set(int i, int x);
    public void Set(int i, string s);
    public void Set(int i, bool b);

    public P Result { get; }
  }
}
```

Using this pattern would allow for flexibility and extensibility, e.g. to construct an `ImmutableArray<T>` or `ReadOnlyCollection<T>` by providing builders for these types. These builders receive the capacity (which could be implemented as a minimum threshold, e.g. for `Span<T>` types where `P` constructs a view over a possible larger backing object), see sequential `Add` or `Set` methods, and can implement `Result` in different ways, e.g. by sealing a mutable instance before handing it back, or by creating a view over some object. If a `Dispose` is present, it's guaranteed to be called after the method that used the `params`-constructed instance of `P` returns. That allows for a `Return` operation.'

As a boundary condition for the existing `params T[]` support, a builder could be constructed that simply performs `new T[capacity]` and has a `Set` methods that stores into the array. `Dispose` is absent, and `Result` returns the array. The functionality to construct the array is simply moved to a library rather than baking `newarr` and `stelem` instructions. While `T[]` support could just continue to use the existing inline code for efficiency reasons, `params P` for any type `P` could go through the pattern, allowing for tight control over allocation patterns inside the library.

Also, the types that can be used in a `params` argument list are simply derived from the types supported on `Add` or `Set`. For example, if a builder for `Dictionary<K, V>` were to have `Add` or `Set` overloads with `KeyValuePair<K, V>` and `ValueTuple<K, V>`, `params` constructing such a dictionary would support a comma-separated argument list of these types. E.g. `MakeRequest(uri, HttpVerbs.Post, ("header1", "value1"), ("header2", "value2"))`. Whether or not that's desirable is a matter of library design, and it's not much different from supporting `Add` in collection initializers (which also picks up on extension methods). The difference here though is that the builder pattern lends itself well to construction of (immutable) objects that the caller doesn't have a reference to ("anonymous" if you like), while collection initializers don't have such a pattern that has an initial capacity or a phase where the result can be sealed.

Back to the initial example with `Use`, the builder type for a `Span<T>` could be a `SpanBuilder<T>` whose `Create` method uses the specified capacity to rent an array. The `Add` or `Set` methods fill in the array, and `Result` returns a `Span<T>` over the array. The call to `Dispose` returns the array to the pool. As such, the lifetime of the array is limited to the duration of the call. On the flip side, the generated code has a sequence of calls, so there's a question about perf (inlining?).

Note that arguments that are poured into a `params` parameter do not have to be uniformly typed, as pointed out above with the dictionary example. This brings me to the analogy for interpolated strings. The `ValueStringBuilder` type shown here is very similar to my builder type `I` (note that `B` is merely an entry point, and could totally be the same type with a static `Create` method on it), with `Add` being the equivalent of the various overloads of `Append`. If an interpolated string were to be treated as the moral equivalent of a heterogeneous argument list strings and interpolations (triples of an expression of type T, an optional alignment of type int, and an optional format string, represented as a struct a la `FormattedValue<T>`), the mechanism could be the same as `params`. The builder would simply see a chain of calls to `Add(string)` and `Add<T>(FormattedValue<T>)` and can construct into a rented `char[]` array, ultimately returning a `string` (or another `Span<char>`-y thing), and finally releasing the `char[]` array upon the call to `Dispose`. E.g.

```csharp
$"A {b} c {d:e} f {g,h} i {j,k:l}"
```

would become

```csharp
using (var __t = ValueStringBuilder.Create(8)) // NB: capacity isn't that useful here
{
  __t.Set(0, "A ");
  __t.Set(1, FormattedValue.Create(b));
  __t.Set(2, " c ");
  __t.Set(3, FormattedValue.Create(g, h));
  __t.Set(4, " i ");
  __t.Set(5, FormattedValue.Create(j, k, l));
}
```

Alternatively, the pattern could be `Create` with the format string (containing the holes) as the "capacity" and `Add` calls for the interpolated expressions (which then don't have to use some `FormattedValue<T>` that carries all information, given it's in the format string already). I guess the difference is between parsing an initially supplied format string and collecting all the substitutions, versus seeing a "stream" of strings and values coming in an append-only way as a sequence of calls. In the case of call sequences, we don't need to represent an array of variants.

Either way, I felt there's some anology between `params` and interpolated strings when looking at it from a builder pattern point of view, which allows for an open-ended set of types that can support in-place `params` construction. Also see https://github.com/bartdesmet/csharplang/blob/ExpressionTypes/proposals/params-builders.md for an earlier write-up on this train of thought.